contracts/samples/VerifyingPaymaster.sol

```diff
@@ -52,26 +59,48 @@ contract VerifyingPaymaster is BasePaymaster {
     /**
      * verify our external signer signed this request.
      * the "paymasterAndData" is expected to be the paymaster and a signature over the entire request params
+     * paymasterAndData[:20] : address(this)
+     * paymasterAndData[20:28] : validUntil
+     * paymasterAndData[28:36] : validAfter
+     * paymasterAndData[36:] : signature
      */
     function validatePaymasterUserOp(UserOperation calldata userOp, bytes32 /*userOpHash*/, uint256 requiredPreFund)
-    external view override returns (bytes memory context, uint256 sigTimeRange) {
+    external override returns (bytes memory context, uint256 sigTimeRange) {
         (requiredPreFund);
 
         bytes32 hash = getHash(userOp);
+        uint256 nonce = senderNonce[userOp.getSender()]++;
         bytes calldata paymasterAndData = userOp.paymasterAndData;
-        uint256 sigLength = paymasterAndData.length - 20;
+        (uint64 validUntil, uint64 validAfter, bytes calldata signature) = parsePaymasterAndData(paymasterAndData);
+        uint256 sigLength = signature.length;
+
         //ECDSA library supports both 64 and 65-byte long signatures.
         // we only "require" it here so that the revert reason on invalid signature will be of "VerifyingPaymaster", and not "ECDSA"
         require(sigLength == 64 || sigLength == 65, "VerifyingPaymaster: invalid signature length in paymasterAndData");
 
         //don't revert on signature failure: return SIG_VALIDATION_FAILED
-        if (verifyingSigner != hash.toEthSignedMessageHash().recover(paymasterAndData[20 :])) {
-            return ("",1);
+        bytes32 digest = _hashTypedDataV4(keccak256(abi.encode(
+            keccak256("ValidatePaymasterUserOp(bytes32 hash,uint64 validUntil,uint64 validAfter,uint256 senderNonce)"),
+            hash,
+            validUntil,
+            validAfter,
+            nonce
+        )));
+        address signer = ECDSA.recover(digest, signature);
+        if (verifyingSigner != signer) {
+            return ("",packSigTimeRange(true,validUntil,validAfter));
         }
 
         //no need for other on-chain validation: entire UserOp should have been checked
         // by the external service prior to signing it.
-        return ("", 0);
+        return ("",packSigTimeRange(false,validUntil,validAfter));
     }
 
+    function parsePaymasterAndData(bytes calldata paymasterAndData) public pure returns(uint64 validUntil, uint64 validAfter, bytes calldata signature) {
+        assembly {
+            validUntil := calldataload(sub(paymasterAndData.offset, 4))
+            validAfter := calldataload(add(paymasterAndData.offset, 4))
```

Comment:
Nice play with offsets... but I still prefer pure solidity, where it doesn't add much overhead..